### PR TITLE
[DI] Add section about getter injection

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Ref https://github.com/symfony/symfony/issues/20973

### CODE REVIEW

File: service_container/injection_types.rst [hunk 1]
```diff
@@ -179,6 +179,88 @@ The disadvantages of setter injection are:
...
+
+* You can be sure that the dependency will not change during the object's
+  lifetime.
+
+* It works well with class hierarchies since you can also override getters of
+  parent classes.
+
+The disadvantages of getter injection are:
+
+* You must call the getter everytime you need the dependency internally.
```
Comment (ogizanagi) [inline service_container/injection_types.rst]: There isn't any property holding the dependency to use anyway. Is it worth mentioning this as a drawback? 😅

File: service_container/injection_types.rst [hunk 2]
```diff
@@ -179,6 +179,88 @@ The disadvantages of setter injection are:
...
+
+    .. code-block:: yaml
+
+       services:
+            # ...
+
+            app.newsletter_manager:
+                class: AppBundle\Mail\NewsletterManager
+                getters:
+                    - getMailer: '@mailer'
```
Comment (ogizanagi) [inline service_container/injection_types.rst]: Shouldn't it rather be:

```yml

     app.newsletter_manager:
         class: AppBundle\Mail\NewsletterManager
         getters:
             getMailer: '@mailer'
             getLogger: '@logger'
```

?

File: service_container/injection_types.rst [hunk 3]
```diff
@@ -179,6 +179,88 @@ The disadvantages of setter injection are:
...
+                    - getLogger: '@logger'
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
```
Comment (ogizanagi) [inline service_container/injection_types.rst]: You may know that, but the schema update is missing from the code PR ^^'

Comment (nicolas-grekas) [inline service_container/injection_types.rst]: I forgot about it, fixed, thanks for the reminder!

File: service_container/injection_types.rst (hunk 2 above)
Comment (nicolas-grekas) [inline service_container/injection_types.rst]: fixed

File: service_container/injection_types.rst (hunk 1 above)
Comment (nicolas-grekas) [inline service_container/injection_types.rst]: That's actually the same drawback as outlined here, reformulated :)

Comment (ogizanagi) [inline service_container/injection_types.rst]: I meant this should probably not appear in the drawback section at all. It isn't one. Just a consequence of the above. Which has no impact.

Comment (nicolas-grekas) [inline service_container/injection_types.rst]: oh! removed then!

File: service_container/injection_types.rst [hunk 4]
```diff
@@ -179,6 +179,86 @@ The disadvantages of setter injection are:
...
+Getter Injection
+----------------
+
+Another possible injection point into a class is by overriding a getter method
+to make it return the dependency::
+
+    // ...
+    abstract class NewsletterManager
+    {
+        abstract protected function getMailer(): MailerInterface;
```
Comment (javiereguiluz) [inline service_container/injection_types.rst]: If these return types are mandatory, we must add somewhere that this only works with PHP 7. Otheriwse, we should remove them.

Comment (nicolas-grekas) [inline service_container/injection_types.rst]: I forgot to mention that PHP 7 is required when using getter injection. Added now.
About the type hints, they are not required, but highly recommended. Given the PHP 7 requirement, I think we should keep them here.

File: service_container/injection_types.rst [hunk 5]
```diff
@@ -179,6 +179,89 @@ The disadvantages of setter injection are:
 * You cannot be sure the setter will be called and so you need to add checks
   that any required dependencies are injected.
 
+Getter Injection
+----------------
+
+    .. versionadded:: 3.3
+        Getter Injection was introduced in Symfony 3.3.
```
Comment (fabpot) [inline service_container/injection_types.rst]: We should mention that this is an experimental feature

File: service_container/injection_types.rst [hunk 6]
```diff
@@ -179,6 +179,89 @@ The disadvantages of setter injection are:
 * You cannot be sure the setter will be called and so you need to add checks
   that any required dependencies are injected.
 
+Getter Injection
+----------------
+
+    .. versionadded:: 3.3
```
Comment (xabbuh) [inline service_container/injection_types.rst]: The leading spaces should be removed.

File: service_container/injection_types.rst [hunk 7]
```diff
@@ -179,6 +179,94 @@ The disadvantages of setter injection are:
...
+
+    .. code-block:: php
+
+        use AppBundle\Mail\NewsletterManager;
+        use Symfony\Component\DependencyInjection\Definition;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        // ...
+        $container->register('app.newsletter_manager', NewsletterManager::class)
+            ->addOverriddenGetter('getMailer', new Reference('mailer'))
```
Comment (ogizanagi) [inline service_container/injection_types.rst]: Should be `setOverriddenGetter` (same below).

**Comment by nicolas-grekas** [comment]
Reverted, one less thing to document for 3.3 :)

**Pull request closed without merging**